File: analysis-service/app/utils/log_formatter.py
from datetime import datetime, timezone
from typing import Any


UNKNOWN_VALUE = "unknown"


def _first_present(log: dict[str, Any], *keys: str) -> Any:
    for key in keys:
        value = log.get(key)
        if value is not None:
            return value

    return None


def _format_timestamp(value: Any) -> str:
    if value is None:
        return "an unknown time"

    if isinstance(value, datetime):
        timestamp = value
    elif isinstance(value, str):
        normalized = value.replace("Z", "+00:00")
        try:
            timestamp = datetime.fromisoformat(normalized)
        except ValueError:
            return value
    else:
        return str(value)

    if timestamp.tzinfo is None:
        timestamp = timestamp.replace(tzinfo=timezone.utc)

    return timestamp.strftime("%I:%M %p on %Y-%m-%d").lstrip("0")


def _format_bool(value: Any) -> str:
    return "yes" if bool(value) else "no"


def format_activity_log(log: dict[str, Any]) -> str:
    user_id = _first_present(log, "user_id", "userId") or UNKNOWN_VALUE
    endpoint = _first_present(log, "endpoint", "path") or UNKNOWN_VALUE
    method = (_first_present(log, "method", "httpMethod") or UNKNOWN_VALUE)
    ip_address = _first_present(log, "ip_address", "ipAddress") or UNKNOWN_VALUE
    user_agent = _first_present(log, "user_agent", "userAgent") or UNKNOWN_VALUE
    status_code = _first_present(log, "status_code", "statusCode") or UNKNOWN_VALUE
    risk_score = _first_present(log, "risk_score", "riskScore")
    is_anomaly = _first_present(log, "is_anomaly", "isAnomaly")
    created_at = _first_present(log, "created_at", "createdAt", "timestamp")

    text = (
        f"User {user_id} accessed {endpoint} using {str(method).upper()} "
        f"at {_format_timestamp(created_at)} from IP {ip_address}. "
        f"The request returned status {status_code} and used user agent {user_agent}."
    )

    if risk_score is not None:
        text += f" Deterministic risk score was {risk_score}."

    if is_anomaly is not None:
        text += f" Marked as anomaly: {_format_bool(is_anomaly)}."

    return text


def format_activity_logs(logs: list[dict[str, Any]]) -> list[str]:
    return [format_activity_log(log) for log in logs]

File: backend/src/utils/logFormatter.js
const UNKNOWN_VALUE = "unknown";

function firstPresent(log, ...keys) {
  for (const key of keys) {
    const value = log[key];
    if (value !== null && value !== undefined) {
      return value;
    }
  }
  return null;
}

function formatTimestamp(value) {
  if (value === null || value === undefined) {
    return "an unknown time";
  }

  const date = value instanceof Date ? value : new Date(value);
  if (Number.isNaN(date.getTime())) {
    return String(value);
  }

  const hours24 = date.getUTCHours();
  const minutes = String(date.getUTCMinutes()).padStart(2, "0");
  const period = hours24 >= 12 ? "PM" : "AM";
  const hours12 = hours24 % 12 === 0 ? 12 : hours24 % 12;
  const year = date.getUTCFullYear();
  const month = String(date.getUTCMonth() + 1).padStart(2, "0");
  const day = String(date.getUTCDate()).padStart(2, "0");

  return `${hours12}:${minutes} ${period} on ${year}-${month}-${day}`;
}

function formatBool(value) {
  return value ? "yes" : "no";
}

function formatActivityLog(log) {
  const userId = firstPresent(log, "user_id", "userId") ?? UNKNOWN_VALUE;
  const endpoint = firstPresent(log, "endpoint", "path") ?? UNKNOWN_VALUE;
  const method = firstPresent(log, "method", "httpMethod") ?? UNKNOWN_VALUE;
  const ipAddress = firstPresent(log, "ip_address", "ipAddress") ?? UNKNOWN_VALUE;
  const userAgent = firstPresent(log, "user_agent", "userAgent") ?? UNKNOWN_VALUE;
  const statusCode = firstPresent(log, "status_code", "statusCode") ?? UNKNOWN_VALUE;
  const riskScore = firstPresent(log, "risk_score", "riskScore");
  const isAnomaly = firstPresent(log, "is_anomaly", "isAnomaly");
  const createdAt = firstPresent(log, "created_at", "createdAt", "timestamp");

  let text =
    `User ${userId} accessed ${endpoint} using ${String(method).toUpperCase()} ` +
    `at ${formatTimestamp(createdAt)} from IP ${ipAddress}. ` +
    `The request returned status ${statusCode} and used user agent ${userAgent}.`;

  if (riskScore !== null && riskScore !== undefined) {
    text += ` Deterministic risk score was ${riskScore}.`;
  }

  if (isAnomaly !== null && isAnomaly !== undefined) {
    text += ` Marked as anomaly: ${formatBool(isAnomaly)}.`;
  }

  return text;
}

function formatActivityLogs(logs) {
  return logs.map(formatActivityLog);
}

module.exports = { formatActivityLog, formatActivityLogs };
